src/flag_gems/ops/instancenorm.py

```diff
@@ -0,0 +1,664 @@
+                running_mean is not None and running_var is not None
+            ), "running_mean and running_var should not both be None"
+            assert (
+                running_mean.shape == running_var.shape and running_mean.shape[0] == C
+            ), f"running_mean and running_var should have shape as {[C,]}"
+            assert (
+                running_mean.dtype == running_var.dtype
+            ), "running_mean and running_var should have the same dtype"
+        if not use_input_stats:
+            assert (
+                has_running_stats
+            ), "Expected running_mean and running_var to be defined when use_input_stats is False"
+
+        # NOTE: when the input is half-precision(either float16 or bfloat16)
+        # these statistical data saved for backward is in single precision
+        acc_type = get_accumulator_dtype(x.dtype)
+        mean = torch.empty(size=(B, C), dtype=acc_type, device=x.device)
+        rstd = torch.empty(size=(B, C), dtype=acc_type, device=x.device)
+
+        with torch.cuda.device(x.device):
+            if use_input_stats:
+                if N <= 128:
+                    TILE_N = triton.next_power_of_2(N)
+                    TILE_M = triton.cdiv(1024, TILE_N)
+                    grid = (triton.cdiv(M, TILE_M), 1, 1)
+                    instance_norm_persistent_kernel_multiline[grid](
+                        x,
+                        y,
+                        weight,
+                        bias,
+                        mean,
+                        rstd,
+                        M,
+                        N,
+                        C,
+                        eps,
+                        TILE_M,
+                        TILE_N,
+                        HAS_WEIGHT_BIAS=has_weight_bias,
+                    )
+                elif N <= 4096:
+                    TILE_N = triton.next_power_of_2(N)
+                    grid = (M, 1, 1)
+                    instance_norm_persistent_kernel[grid](
+                        x,
+                        y,
+                        weight,
+                        bias,
+                        mean,
+                        rstd,
+                        M,
+                        N,
+                        C,
+                        eps,
+                        TILE_N,
+                        HAS_WEIGHT_BIAS=has_weight_bias,
+                    )
+                else:
+                    grid = (M, 1, 1)
+                    instance_norm_loop_kernel[grid](
+                        x,
+                        y,
+                        weight,
+                        bias,
+                        mean,
+                        rstd,
+                        M,
+                        N,
+                        C,
+                        eps,
+                        HAS_WEIGHT_BIAS=has_weight_bias,
+                    )
+                if has_running_stats and use_input_stats:  # update running stats
```

Comment:
redundant condition

Reply:
Actually No.
`has_running_stats` means `running_mean` and `running_var` is not None.
`use_input_stats` means use `mean` and `rstd` from input, otherwise uses `running_mean` and `running_var` from outside.
Condition when updating running stats, is `has_running_stats` and `use_input_stats` be True simultaneously.

Reply:
use_input_stats is checked in line 500. so there is no need to check it again.